gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/jvmkill.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <sys/types.h>
+#include <signal.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include <jvmti.h>
+
+static void JNICALL
+resourceExhausted(
+      jvmtiEnv *jvmti_env,
+      JNIEnv *jni_env,
+      jint flags,
+      const void *reserved,
+      const char *description)
+{
+   fprintf(stderr,
+      "ResourceExhausted: %s: killing current process!\n", description);
+   kill(getpid(), SIGKILL);
+}
+
+JNIEXPORT jint JNICALL
+Agent_OnLoad(JavaVM *vm, char *options, void *reserved)
+{
+   jvmtiEnv *jvmti;
+   jvmtiError err;
+
+   jint rc = (*vm)->GetEnv(vm, (void **) &jvmti, JVMTI_VERSION);
+   if (rc != JNI_OK) {
+      fprintf(stderr, "ERROR: GetEnv failed: %d\n", rc);
```

Comment:
Where do these fprintf's go at the time of crash? Normally these are redirected to "start_snappy..." logs but perhaps will be better to send them to a new file instead (like using mkstemp) because the JVM is OOM and could fail in writing to start_ logs too. Also as discussed we could add dumping using jmap when CRITICAL_UP is reached.